### PR TITLE
Data Acquisition from Board

### PR DESCRIPTION
**Pull request by metavore**
### 1. Resource Management:

 - Context manager ensures proper cleanup of board resources
 - Automatic stream stop/release even if errors occur
 - RAII principles followed through context manager pattern

### 2. Thread Safety:

 - Shared lock protects board operations across threads
 - Safe concurrent access to board data
 - Protected cleanup operations

### 3. Processing Features:

- Configurable window sizes and overlap
- Non-blocking data acquisition
- Separate processing thread with proper error handling
- Flexible callback interface for data processing

### 4. Code Quality:

- Minimal wrapping of BrainFlow's BoardShim
- Simple example usage provided
- Type hints throughout
- Clean separation of concerns between streaming and processing

### 5. Design Benefits:

- No unnecessary abstractions over BoardShim
- Direct access to board when needed
- Flexible processing configuration
- Easy to integrate with existing code

### CODE REVIEW

File: neuros/board/stream_processor.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import threading
+import time
+from brainflow.board_shim import BoardShim, BrainFlowInputParams, BoardIds
+
+# Global lock for thread-safe board operations
+_board_lock = threading.Lock()
+
+
+@contextmanager
+def board_stream(board_id: int = BoardIds.SYNTHETIC_BOARD) -> Iterator[BoardShim]:
```
Comment (EricHennigan) [inline neuros/board/stream_processor.py]: Pretty sure that it's not returning an Iterator, but even if it is, that's the wrong data type. We don't have a stream of boards, we have one board with a stream of data

File: neuros/board/stream_processor.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        self.board = board
+        self.config = config or ProcessorConfig()
+
+        # Convert ms to samples
+        sample_rate = BoardShim.get_sampling_rate(board.get_board_id())
+        self.window_samples = int(sample_rate * self.config.window_size_ms / 1000)
+        self.overlap_samples = int(sample_rate * self.config.overlap_ms / 1000)
+
+        # Processing thread management
+        self._process_thread: Optional[threading.Thread] = None
```
Comment (EricHennigan) [inline neuros/board/stream_processor.py]: Why do some of the fields have type annotations and others don't?

File: neuros/board/stream_processor.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        if self._process_thread is not None:
+            self._process_thread.join()
+            self._process_thread = None
+
+    def _process_loop(self) -> None:
+        """Main processing loop."""
+        while self._running:
+            try:
+                # Get latest data window - thread-safe
+                with _board_lock:
```
Comment (EricHennigan) [inline neuros/board/stream_processor.py]: Why does reading data from the board need a lock?

File: neuros/board/stream_processor.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+
+@dataclass
+class ProcessorConfig:
+    """Configuration for the stream processor"""
+    window_size_ms: float = 50.0  # Size of processing window in milliseconds
+    overlap_ms: float = 0.0  # Overlap between windows in milliseconds
+
+
+class StreamProcessor:
```
Comment (EricHennigan) [inline neuros/board/stream_processor.py]: How many of these am I allowed to create and point at the same board?

File: neuros/board/stream_processor.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        self._process_thread.start()
+
+    def stop(self) -> None:
+        """Stop processing and clean up."""
+        self._running = False
+        if self._process_thread is not None:
+            self._process_thread.join()
+            self._process_thread = None
+
+    def _process_loop(self) -> None:
```
Comment (EricHennigan) [inline neuros/board/stream_processor.py]: This method is meant to run as a background thread, but it refers to variables held by the current instance created on the main thread. For example, the _on_data field. Can we avoid having memory shared across threads?

File: neuros/board/stream_processor.py (hunk 1 above)
Comment (metavore) [reply neuros/board/stream_processor.py]: You are correct. It should return a BoardShim. Fixing.

File: neuros/board/stream_processor.py (hunk 2 above)
Comment (metavore) [reply neuros/board/stream_processor.py]: I started rushing. Consistency incoming!

File: neuros/board/stream_processor.py (hunk 3 above)
Comment (metavore) [reply neuros/board/stream_processor.py]: This was a confused attempt to address the potential memory leak issue, didn't document well enough and confused myself and then rushed to PR. Sorry, cleaning up as part of solution to that other problem.

Comment (metavore) [reply neuros/board/stream_processor.py]: Yes, there's already a separate thread from BrainFlow that starts with the board.start_session() call, good enough for reading.

File: neuros/board/stream_processor.py (hunk 4 above)
Comment (metavore) [reply neuros/board/stream_processor.py]: Currently, there's no limit, and that's bad. I'll freeze it.

File: neuros/board/stream_processor.py (hunk 5 above)
Comment (metavore) [reply neuros/board/stream_processor.py]: This is a big bad OOPS. A deadly cyber-sin, I say! Careless oversight falls under Sloth. Which level of hell is that one? Will fix it up. I just defaulted to "Need concurrency, make thread, keep it classy" and got distracted by the quest for coffee.

I can rectify this in a couple of ways. We can keep the separate thread, and share nothing - except for one thread.Event() for stopping. Or I can get rid of the class and get rid of the thread and just manage the loop explicitly. Actually, yes, that sounds like the best idea. Classlessness.

File: neuros/main.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    config = WindowConfig(window_ms=500.0, overlap_ms=250.0)
+
+    try:
+        with board_stream() as board:
+            logger.info(f"Board ready - sample rate: {board.get_sampling_rate(board_id=BoardIds.SYNTHETIC_BOARD)} Hz")
+            logger.info("Press Ctrl+C to stop...")
+
+            # Process windows until interrupted
+            for i, window in enumerate(stream_windows(board, config)):
+                channels: int
```
Comment (EricHennigan) [inline neuros/main.py]: remove these 2 lines, type annotation is not needed here

File: neuros/main.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+            try:
+                # Get new data
+                new_data = board.get_current_board_data(samples_needed)
+                if new_data.size == 0:
+                    # No new data available yet
+                    continue
+
+                # Add to buffer
+                buffer = (np.concatenate([buffer, new_data], axis=1) if buffer.size > 0 else new_data)
```
Comment (EricHennigan) [inline neuros/main.py]: It looks like you want a ring buffer

File: neuros/main.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        try:
+            if board.is_prepared():
+                board.stop_stream()
+                board.release_session()
+        except Exception as e:
+            logger.error(f"Error during board cleanup: {str(e)}")
+            # Don't re-raise during cleanup
+
+
+def stream_windows(board: BoardShim, config: WindowConfig) -> Iterator[np.ndarray]:
```
Comment (EricHennigan) [inline neuros/main.py]: Why is it useful to have a sequence of windows?

File: neuros/main.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    # Buffer for overlap samples
+    buffer = np.array([])
+
+    # Placeholder for new data
+    new_data = np.array([])
+
+    try:
+        while True:
+            # Calculate how many new samples we need
+            if buffer.size == 0:
```
Comment (EricHennigan) [inline neuros/main.py]: Why does the buffer.size influence subtracting overlap_samples?

File: neuros/main.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+                buffer = (np.concatenate([buffer, new_data], axis=1) if buffer.size > 0 else new_data)
+
+                # Once we have enough samples, yield a window
+                if buffer.shape[1] >= window_samples:
+                    # Extract the window
+                    window = buffer[:, :window_samples]
+                    yield window
+
+                    # Update buffer for next iteration
+                    if overlap_samples > 0:
```
Comment (EricHennigan) [inline neuros/main.py]: Pretty sure you can avoid the check entirely with `buffer[:, len(buffer)-overlap_samples:]`

File: neuros/main.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+                        buffer = buffer[:, -overlap_samples:]
+                    else:
+                        buffer = np.array([])
+
+            except Exception as e:
+                logger.error(f"Error getting board data: {str(e)}")
```
Comment (EricHennigan) [inline neuros/main.py]: Pretty sure that the str() conversion happens automatically and you don't need to make it explicit

File: neuros/main.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+        logger.info("Stopping window streaming")
+
+        # Add to buffer
+        buffer = (np.concatenate([buffer, new_data], axis=1) if buffer.size > 0 else new_data)
+
+        # Once we have enough samples, yield a window
+        if buffer.shape[1] >= window_samples:
+            # Extract the window
+            window = buffer[:, :window_samples]
+            yield window
```
Comment (EricHennigan) [inline neuros/main.py]: Why is there a yield in the GeneratorExit exception?

File: neuros/main.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    window_samples, overlap_samples = config.to_samples(sample_rate)
+
+    # Buffer for overlap samples
+    buffer = np.array([])
+
+    # Placeholder for new data
+    new_data = np.array([])
+
+    try:
+        while True:
```
Comment (EricHennigan) [inline neuros/main.py]: Should the loop be caller-side?

File: neuros/main.py (hunk 6 above)
Comment (metavore) [reply neuros/main.py]: Dah. Duh. Done.

File: neuros/main.py (hunk 7 above)
Comment (metavore) [reply neuros/main.py]: This is a new concept for me, but after learning a bit more, yeah, that would be simpler.

File: neuros/main.py (hunk 8 above)
Comment (metavore) [reply neuros/main.py]: We need fixed-size chunks for FFT, but we also don't want to miss frequencies that are literally "edge cases". And on top of that, we want continuous processing - with neurofeedback I think that any millisecond pause would be unacceptable, so it's also insurance.

"Slow is smooth, smooth is fast."

However, this gave me a thought, which I'll share below at the last reply.

File: neuros/main.py (hunk 9 above)
Comment (metavore) [reply neuros/main.py]: I think that was a lazy temporary fix for the first-window problem. The synthetic board doesn't start instantaneously and could be totally empty in the first window.

File: neuros/main.py (hunk 10 above)
Comment (metavore) [reply neuros/main.py]: That didn't occur to me, but makes sense.

File: neuros/main.py (hunk 11 above)
Comment (metavore) [reply neuros/main.py]: I knew this! That was me letting the LLM double-check my work, without double-checking its work. My face, my palm, they meet.

File: neuros/main.py (hunk 12 above)
Comment (metavore) [reply neuros/main.py]: That makes no sense. I think it was a misplaced code fragment. In any case it's accidental and silly.

File: neuros/main.py (hunk 13 above)
Comment (metavore) [reply neuros/main.py]: Opinion: That doesn't change anything functionally, and the caller doesn't really gain any advantage, just an extra job. The current way also makes it visually clearer that we are necessarily working with a stream of data.

File: neuros/main.py [hunk 14]
```diff
@@ -47,125 +46,71 @@ def board_stream(board_id: int = BoardIds.SYNTHETIC_BOARD,
...
-                    window = buffer[:, :window_samples]
-                    yield window
-
-                    # Update buffer for next iteration
-                    if overlap_samples > 0:
-                        buffer = buffer[:, -overlap_samples:]
-                    else:
-                        buffer = np.array([])
+                ring_buffer[:, :-new_samples] = ring_buffer[:, new_samples:]
+                ring_buffer[:, -new_samples:] = eeg_data
```
Comment (EricHennigan) [inline neuros/main.py]: I don't think this code works as you'd expect.

Some things to check:
  - is new_data always equal to the new_samples in size?
  - what happens to the data in-between? I mean, the first line copies the tail end of the ring_buffer to the beginning. But then the next like copies eeg_data into the tail end vs where the new_samples ended. Perhaps you meant to leave out the minus sign?
  - would a test reveal an error in this copying?

**Comment by metavore** [comment]
I'm not sure if the PR description is better like that, or like this:

Data Acquisition: Initial Window Generation Implementation

This PR introduces the core data acquisition pipeline, providing a clean and 
efficient foundation for streaming windowed EEG data.

Key Features:
- Functional generator-based approach for window streaming
- Efficient window management with configurable overlap
- Thread-safe board operations with proper resource management
- Comprehensive test suite using Given-When-Then pattern
- Clear separation of core functionality and usage examples

Technical Details:
- WindowConfig dataclass for validated configuration
- Context manager pattern for board lifecycle
- Generator-based window stream implementation
- Accumulator-based continuous data handling with proper overlap
- Thorough error handling and cleanup
- Complete test coverage of all core functionality

All tests passing. Ready for signal processing integration.

File: tests/test_window_streams.py [hunk 15]
```diff
@@ -3,160 +3,182 @@
...
-        window = next(generator)
-
-        assert isinstance(window, np.ndarray)
-        assert window.shape == (len(eeg_channels), expected_samples)
+        assert isinstance(window, np.ndarray), "Window should be numpy array"
+        assert window.shape == (len(eeg_channels), expected_samples), "Window shape incorrect"
+        assert not np.any(np.isnan(window)), "Window should not contain NaN values"
 
 
 def test_window_overlap():
```
Comment (EricHennigan) [inline tests/test_window_streams.py:69]: Can we add a test where the window_ms, frequency of sampling, and overlap_ms share no common factors?

Comment (metavore) [reply tests/test_window_streams.py:69]: We certainly can.

**Review by EricHennigan — APPROVED** [review]

**Pull request merged**